awscrt/http.py

```diff
@@ -100,92 +109,147 @@ def on_connection_setup(binding, error_code):
-    of the request.
 
-    on_read_body is invoked to read the body of the request. It takes a single parameter of type MemoryView
-    (it's writable), and you signal the end of the stream by returning OutgoingHttpBodyState.Done for the first tuple
-    argument. If you aren't done sending the body, the first tuple argument should be OutgoingHttpBodyState.InProgress
-    The second tuple argument is the size of the data written to the memoryview.
+    def _on_response(self, status_code, name_value_pairs):
+        self.response_status_code = status_code
+
+        headers = HttpHeaders()
+        for pair in name_value_pairs:
+            headers.add(pair[0], pair[1])
+
+        if self._on_response_cb:
+            self._on_response_cb(self, status_code, headers)
 
-    on_incoming_body is invoked as the response body is received. It takes a single argument of type bytes.
+
+class HttpMessageBase(NativeResource):
+    """
+    Base for HttpRequest and HttpResponse classes.
     """
-    __slots__ = ('_connection', 'path_and_query', 'method', 'outgoing_headers', '_on_read_body', '_on_incoming_body', '_stream',
-                 'response_headers', 'response_code', 'has_response_body', 'response_headers_received',
-                 'response_completed')
+    __slots__ = ('headers', '_body_stream')
 
-    def __init__(self, connection, method, path_and_query, outgoing_headers, on_read_body, on_incoming_body):
-        assert method is not None
-        assert outgoing_headers is not None
-        assert connection is not None and isinstance(connection, HttpClientConnection)
+    def __init__(self, body_stream=None):
+        assert isinstance(body_stream, IOBase) or body_stream is None
 
-        self.path_and_query = path_and_query
+        super(HttpMessageBase, self).__init__()
+        self.headers = HttpHeaders()
+        self._body_stream = body_stream
 
-        if path_and_query is None:
-            self.path_and_query = '/'
+    @property
+    def body_stream(self):
+        return self._body_stream
 
-        self._connection = connection
+class HttpRequest(HttpMessageBase):
+    """
+    Definition for an outgoing HTTP request.
+    The request may be transformed (ex: signing the request) before its data is eventually sent.
+    """
+
+    __slots__ = ('method', 'path')
+
+    def __init__(self, method='GET', path='/', body_stream=None):
+        super(HttpRequest, self).__init__(body_stream)
         self.method = method
-        self.outgoing_headers = outgoing_headers
-        self._on_read_body = on_read_body
-        self._on_incoming_body = on_incoming_body
-        self.response_completed = Future()
-        self.response_headers_received = Future()
-        self._stream = None
-        self.response_headers = None
-        self.response_code = None
-        self.has_response_body = False
+        self.path = path
+        self._binding = _awscrt.http_request_new(self, self._body_stream)
+
+
+class HttpHeaders(object):
```

Comment:
The idea was that users could always access the `HttpHeaders.map` for a normal Pythonic experience, and the functions on this class were just helpers for common operations. This was inspired by [golang's HttpHeader type](https://golang.org/pkg/net/http/#Header) which is just a map with extra functions attached it.

I could learn more about making something natively iterable in python. That would open the door to binding this more tightly to C datastrcture, instead of lazily updating the C datastructure...

Reply:
iterable
and so much more...

Reply:
map isn't public tho?